### ISSUE TITLE
Quoted booleans are being allowed even with `isLenient = false`

### ISSUE THREAD

**Issue by rewgoes** [Kotlin/kotlinx.serialization#2749]
**Describe the bug**
I have `isLenient = false` but the serializer is allowing quoted booleans, which I believe goes against [RFC-4627](https://www.ietf.org/rfc/rfc4627.txt).

Example:
```json
{
 "aBoolean": "true"
}
```

The JSON above should not be accepted based on RFC-4627.

The issue was introduces by #2440.

**To Reproduce**
User the following JSON configuration and try to parse a quoted boolean as mentioned above:
```kotlin
Json {
  isLenient = false
}
```

**Expected behavior**
The parser should fail and throw an exception:  `JsonConvertException`

**Environment**
 - Kotlin version: 1.9.20
 - Library version: 1.6.3

**Comment by sandwwraith** [comment]
RFC does not say anything about `"true"` — if we talk purely about the Json standard, it is just a string value with 4 characters and not a boolean value. This problem and all similar ones stem from the fact that there is no RFC, standard, or anything else on how Json values should be mapped onto particular programming language types, so it is mostly up for consideration of a particular library's implementation. Given the fluid nature of Json, web APIs, standards, and everything else, I think it is quite handy if both `true` and `"true"` can be mapped onto `val aBoolean: Boolean` and `val s: String` — since often Json producers are not in control of Json consumers. If you think this may cause problems, please write about your use case.

**Comment by rewgoes** [comment]
RFC online define the usage of quotation marks for strings, not booleans:
> string = quotation-mark *char quotation-mark

If someone intends to accept quoted booleans, they should define `isLenient = true`, which basically bypass any RFC
> Removes JSON specification restriction (RFC-4627) and makes parser more liberal to the malformed input. In lenient mode, unquoted JSON keys and string values are allowed.


About your comment here:
> there is no RFC, standard, or anything else on how Json values should be mapped onto particular programming language types, so it is mostly up for consideration of a particular library's implementation

I beg to differ on this, the RFC is pretty clear on what everything should be mapped to:
> JSON can represent four primitive types (strings, numbers, booleans, and null) and two structured types (objects and arrays).

And the possible values to be:
```
   The literal names MUST be lowercase.  No other literal names are
   allowed.

         value = false / null / true / object / array / number / string

         false = %x66.61.6c.73.65   ; false

         null  = %x6e.75.6c.6c      ; null

         true  = %x74.72.75.65      ; true
```

If we map what is clearly a string to boolean, simply because the value is `"true"` or `"false"`, the parser is being partially lenient, even when the client disabled it (`isLenient = false`).

However, looking back at the reason this change was done, #2438:
> The primitives are allowed to be quoted and unquoted to simplify map key parsing and integrations with third-party API.

If `isLenient = false`, couldn't we simply allow quoted primitive types when parsing map keys, instead of doing values? 

If not that, as long as the library allows primitive types to be quotes, I would simply remove any reference of `RFC-4627` from the `isLenient`, as it is not 100% conform to RFC. Or make it clear in the documentation that quoted primitive types are allowed, or better yet, add another flag to control that (potentially still allowing it quoted primitives on map keys.)

> If you think this may cause problems, please write about your use case.
Fortunately, I had tests for exactly that when I updated the library in my project. But yes, changes like this can cause problems and go unnoticed to production of any app.

**Comment by rewgoes** [comment]
I understand your point and agree that the RFC does not specify how data should be mapped, leaving each language to determine its own method for parsing JSON within its limitations. For example, in C (pre-99), a `boolean` would likely be mapped to `0` or `1`.

However, I must emphasize that the RFC explicitly defines what constitutes a `string` and a `boolean`:

> A string begins and ends with quotation marks

and

> A JSON value MUST be an object, array, number, or string, or one of the following three literal names:
>      `false` `null` `true`

Therefore, if a JSON field contains the value `"true"`, **regardless of the language that parses it**, this is a `string` as per the RFC's definition.

Given this, any language or library must represent these types accurately and without ambiguity, according to the RFC.

Additionally, the RFC is based on the ECMAScript Programming Language Standard, Third Edition [ECMA], which states:
> A boolean value is a member of the type Boolean and is one of two unique values, true and false.

Source: [ECMA-262](https://ecma-international.org/wp-content/uploads/ECMA-262_3rd_edition_december_1999.pdf)

If further validation is needed, you can consult Gemini, ChatGPT, Copilot, or any other chatbot with the question:
> "According to https://www.ietf.org/rfc/rfc4627.txt, can a boolean be quoted?"

While chatbots may not always be perfect for programming tasks, they generally excel at understanding and summarizing documentation.

Lastly, though not directly related, most programming languages, including Kotlin, would interpret a quoted boolean as a string:
```kotlin
val stringWithIdentityCrisis : Boolean = "true" // Error: Type mismatch

// or
val aRealString = "true"

if (aRealString) { } // Error: Type mismatch
```

In conclusion, I believe I have made my point clear and do not wish to prolong this discussion. However, if you need assistance in maintaining the `isLenient` flag according to RFC and ECMA standards, while also enabling the library to parse `Map<Int, Int>` and `Map<Boolean, Boolean>` with `isLenient = false`, I would be happy to help with the necessary code changes.

**Comment by rewgoes** [comment]
Just a suggestion, I think it's worth to add this to the release notes for the [version](https://github.com/Kotlin/kotlinx.serialization/releases/tag/v1.6.1) where this was changed. Maybe by saying:
> Allow string literal `"true"` and `"false"` to be parsed as `Boolean`

As I've mentioned, in my opinion, the chances of this change causing any problem in a real app is very rare, and I only identified that because I had some unit test code when I was switching from GSON to this library, to be more confident during the transition. Testing each type in isolation.

**Comment by JakeWharton** [comment]
> Therefore, if a JSON field contains the value `"true"`, **regardless of the language that parses it**, this is a `string` as per the RFC's definition.

This is a string value in the encoded form, yes. The same as unquoted `true` is a boolean in the encoded form.

> Given this, any language or library must represent these types accurately and without ambiguity, according to the RFC.

I'm not sure why you're making this jump. The RFC says nothing about how a language or library should represent the types. It concerns itself with the serialized form only and its semantics.

No one is questioning that `"true"` parses as a JSON string and `true` parses as a JSON boolean per the RFC. If you ask the library for the parsed JSON model you will get a `JsonString` and `JsonBoolean`, respectively, in fact.

However, when doing object binding, a concept which the RFC makes no mention, the library has freedom to do whatever it wants. We can discuss whether the behavior makes sense, but the RFC doesn't help us here.

In general I'm opp
